src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/UserRoleServiceTests.cs

```diff
@@ -552,105 +933,358 @@ await using (var contentDbContext = InMemoryApplicationDbContext(contentDbContex
-            };
+                await userAndRolesDbContext.AddAsync(user);
+                await userAndRolesDbContext.SaveChangesAsync();
 
-            var userRole1 = new IdentityUserRole<string>
-            {
-                UserId = user.Id,
-                RoleId = role1.Id
-            };
+                await contentDbContext.Releases.AddAsync(release);
+                await contentDbContext.SaveChangesAsync();
+            }
+
+            var emailTemplateService = new Mock<IEmailTemplateService>(Strict);
+
+            emailTemplateService.Setup(mock =>
+                    mock.SendReleaseRoleEmail(user.Email,
+                        It.Is<Release>(r => r.Id == release.Id),
+                        Contributor))
+                .Returns(Unit.Instance);
+
+            var userManager = MockUserManager();
+
+            // Here we are checking to see that the user is "upgraded" to the higher-powered Analyst role by their
+            // existing PrereleaseUser role being removed and the higher-powered Analyst role being added. 
+            userManager
+                .Setup(s => s.GetRolesAsync(ItIsUser(user)))
+                .ReturnsAsync(ListOf(RoleNames.PrereleaseUser));
+            
+            userManager
+                .Setup(s => s.AddToRoleAsync(ItIsUser(user), RoleNames.Analyst))
+                .ReturnsAsync(new IdentityResult());
+
+            userManager
+                .Setup(s => s.RemoveFromRolesAsync(ItIsUser(user), 
+                        ItIs.ListSequenceEqualTo(ListOf(RoleNames.PrereleaseUser))))
+                .ReturnsAsync(new IdentityResult());
+
+            await using (var userAndRolesDbContext = InMemoryUserAndRolesDbContext(userAndRolesDbContextId))
+            await using (var contentDbContext = InMemoryApplicationDbContext(contentDbContextId))
+            {
+                var service = SetupUserRoleService(
+                    usersAndRolesDbContext: userAndRolesDbContext,
+                    contentDbContext: contentDbContext,
+                    emailTemplateService: emailTemplateService.Object,
+                    userManager: userManager.Object);
+
+                var result = await service.AddReleaseRole(userId, release.Id, Contributor);
+                
+                VerifyAllMocks(emailTemplateService, userManager);
+                
+                emailTemplateService.Verify(mock =>
+                        mock.SendReleaseRoleEmail(user.Email,
+                            It.Is<Release>(p => p.Id == release.Id),
+                            Contributor),
+                    Times.Once);
+
+                result.AssertRight();
+            }
+
+            await using (var contentDbContext = InMemoryApplicationDbContext(contentDbContextId))
+            {
+                var userReleaseRoles = await contentDbContext.UserReleaseRoles
+                    .AsQueryable()
+                    .ToListAsync();
+                Assert.Single(userReleaseRoles);
```

Comment:
You could `var assignedReleaseRole = Assert.Single(...` here instead of referring to `userReleaseRoles[0]` below. Looks like there's an opportunity to do this elsewhere in other changes throughout this class too.